server/src/main/java/org/opensearch/search/SearchService.java

```diff
@@ -247,6 +250,13 @@ public class SearchService extends AbstractLifecycleComponent implements IndexEv
         Property.NodeScope
     );
 
+    public static final Setting<Boolean> CLUSTER_CONCURRENT_SEGMENT_SEARCH_SETTING = Setting.boolSetting(
```

Comment:
I am wondering if we really need a cluster setting as well, it really conflicts with the respective  index settings: which one takes precedence?

Reply:
The index setting takes precedence, I tried to make this more clear with the truth table I added to the new UT:

https://github.com/opensearch-project/OpenSearch/blob/a412d570049bfcd31dc6154c29e821e00e93e1b3/server/src/test/java/org/opensearch/search/SearchServiceTests.java#L1185-L1197

From a user perspective it would be useful to be able to override the default behavior for all indices, including any newly created indices. For example, currently the default for concurrent search is false. If a user wants to default to use concurrent search for all indices, if the cluster setting doesn't exist then they need to apply the index setting to all existing indices (easy enough), but then also be sure to include the index setting in any new indices that get created.

Reply:
Users won't be looking into IT, thinking from user perspective, it is difficult to navigate between 3 (!) different handles for the same feature. Personally, I think adding index settings (+ feature flag) is a good start, we could add cluster setting later on if there will be a need to do that.

Reply:
Taking a step back though, do you think the default for the index setting should be true then? Today if a user enables the concurrent search feature flag then concurrent search will be enabled. But after introducing the index setting if it defaults to false then the behavior would change and that could lead to confusion.

I'm okay with waiting until when we remove the feature flag to make a decision on if we should replace it with a dynamic cluster setting or leave it as is but right now I'm thinking it's actually less confusing if we keep the dynamic cluster setting but have it default to try to keep the default behavior the same.

Tagging @sohami for any thoughts on this too

Reply:
> Taking a step back though, do you think the default for the index setting should be true then? Today if a user enables the concurrent search feature flag then concurrent search will be enabled. But after introducing the index setting if it defaults to false then the behavior would change and that could lead to confusion.

This is a good point, right now it is all in or nothing (with respect to concurrent search). We want users who enabled this experimental feature not to be impacted by newly introduced index setting (it should work as before). In this regards the default should be `true` indeed.

Reply:
`FeatureFlag` is a temporary setting which will be deprecated with GA. We can treat `FeatureFlag` as a mechanism to initialize `ConcurrentQueryPhaseSearcher`. Then cluster & index settings will be used to dynamically control the behavior to use concurrent vs non concurrent path.

#### Before GA:
1. Users can enable FeatureFlag, then they should be able to use concurrent search. For this can we default cluster setting with the Feature Flag value ? Index setting can still be `false` as default. With this users will not have to change any cluster/index setting in addition to feature flag.
2. Users wants to now dynamically switch to non concurrent path for all indices, they can use cluster setting to disable it even though `ConcurrentQueryPhaseSearcher` is initialized
3. Users wants to now dynamically enable concurrent/non-concurrent path for some indices, they can use index setting to enable/disable it on those indices.

#### Post GA:
1. Feature Flag will be deprecated and cluster setting and index setting will be defaulted to false.
2. User want to enable it globally they can do so by using the cluster setting
3. User want to selectively enable/disable for certain indices where index settings will be useful.